### PR TITLE
Updated Program model with MicroMasters fields

### PR DESCRIPTION
**Pull request by clintonb**
ECOM-5094

### CODE REVIEW

**Comment by clintonb** [comment]
@edx/ecommerce (esp. @schenedx) these are the fields necessary to power a MicroMasters detail page. Please review.

**Comment by schenedx** [comment]
LGTM 👍

**Comment by clintonb** [comment]
@edx/ecommerce please review

File: course_discovery/apps/api/serializers.py [hunk 1]
```diff
@@ -50,8 +50,8 @@
 }
 
 PROGRAM_SEARCH_FIELDS = (
-    'text', 'uuid', 'title', 'subtitle', 'category', 'marketing_url', 'organizations', 'content_type', 'image_url',
-    'status',
+    'text', 'uuid', 'title', 'subtitle', 'category', 'marketing_url', 'authoring_organizations', 'content_type',
+    'image_url', 'status',
```
Comment (schenedx) [inline course_discovery/apps/api/serializers.py]: Do we have 'image_url' in `program` model? If not, maybe we should remove this?

Comment (schenedx) [inline course_discovery/apps/api/serializers.py]: Or rename this to `card_image_url` like you did on line 267?

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+    class Meta:
+        verbose_name = _('FAQ')
+        verbose_name_plural = _('FAQs')
 
-    image = models.ForeignKey(Image, default=None, null=True, blank=True)
 
-    organizations = models.ManyToManyField(Organization, blank=True)
+class ProgramType(TimeStampedModel):
+    name = models.CharField(max_length=32, unique=True, null=False, blank=False)
+    applicable_seat_types = models.ManyToManyField(
```
Comment (schenedx) [inline course_discovery/apps/course_metadata/models.py:500]: Just want to ask a design question: Is this list of seat types required? Or Optional? If this list is empty for MM, does that mean that MM allows course runs of all seat types? Or No seat type is allowed? Maybe some comments about the usage of this data would be good here?

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -516,11 +550,52 @@ def marketing_url(self):
         return None
 
     @property
-    def image_url(self):
-        if self.image:
-            return self.image.src
+    def course_runs(self):
```
Comment (schenedx) [inline course_discovery/apps/course_metadata/models.py:564]: Just a heads up: I believe the existing programs API organizes course_runs into courses. This means, in order for me to return the same json format, I would have to figure out the course set with course_runs already filtered. This method do not get me the list of courses filtered. I will need to create my own method somehow.

File: course_discovery/apps/course_metadata/search_indexes.py [hunk 4]
```diff
@@ -114,10 +114,17 @@ class ProgramIndex(OrganizationsMixin, BaseIndex, indexes.Indexable):
     subtitle = indexes.CharField(model_attr='subtitle')
     category = indexes.CharField(model_attr='category', faceted=True)
     marketing_url = indexes.CharField(null=True)
-    organizations = indexes.MultiValueField(faceted=True)
-    image_url = indexes.CharField(model_attr='image_url', null=True)
+    authoring_organizations = indexes.MultiValueField(faceted=True)
+    credit_backing_organizations = indexes.MultiValueField(faceted=True)
+    image_url = indexes.CharField(model_attr='card_image_url', null=True)
```
Comment (schenedx) [inline course_discovery/apps/course_metadata/search_indexes.py]: Any reason to keep the `image_url` index here?

File: course_discovery/apps/course_metadata/models.py [hunk 5]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+    )
 
+
+class Program(TimeStampedModel):
+    uuid = models.UUIDField(blank=True, default=uuid4, editable=False, unique=True, verbose_name=_('UUID'))
+    title = models.CharField(
+        help_text=_('The user-facing display title for this Program.'), max_length=255, unique=True)
+    subtitle = models.CharField(
+        help_text=_('A brief, descriptive subtitle for the Program.'), max_length=255, blank=True)
+    # TODO Remove category in favor of type
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py:519]: Will you be addressing TODOs like these prior to merging this PR? This one and the SeatType FK one above seem particularly important.

File: course_discovery/apps/course_metadata/models.py [hunk 6]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+    category = models.CharField(help_text=_('The category / type of Program.'), max_length=32)
+    type = models.ForeignKey(ProgramType, null=True, blank=True)
+    status = models.CharField(help_text=_('The lifecycle status of this Program.'), max_length=24)
+    marketing_slug = models.CharField(
+        help_text=_('Slug used to generate links to the marketing site'), blank=True, max_length=255, db_index=True)
+    courses = models.ManyToManyField(Course)
+    # NOTE (CCB): Editors of this field should validate the values to ensure only CourseRuns associated
+    # with related Courses are stored.
+    excluded_course_runs = models.ManyToManyField(CourseRun)
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py:531]: I can see a serializer using this field to decide what CourseRuns to filter out of a response. If someone adds a CourseRun which isn't associated with related Courses, it's going to be end up excluded regardless. Is the validation described in your comment really important?

**Comment by schenedx** [comment]
@clintonb Some minor comments and question. Overall it looks good.
Let me confirm again, you are going to use a different PR for the migration for program-course_run linkage, correct? Also, correctly, you are only importing the url of 1 banner image size. You will import more once the banner images are done correctly, right?

File: course_discovery/apps/course_metadata/models.py [hunk 7]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+class Program(TimeStampedModel):
+    uuid = models.UUIDField(blank=True, default=uuid4, editable=False, unique=True, verbose_name=_('UUID'))
+    title = models.CharField(
+        help_text=_('The user-facing display title for this Program.'), max_length=255, unique=True)
+    subtitle = models.CharField(
+        help_text=_('A brief, descriptive subtitle for the Program.'), max_length=255, blank=True)
+    # TODO Remove category in favor of type
+    category = models.CharField(help_text=_('The category / type of Program.'), max_length=32)
+    type = models.ForeignKey(ProgramType, null=True, blank=True)
+    status = models.CharField(help_text=_('The lifecycle status of this Program.'), max_length=24)
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py]: I'd restrict this to a set of status choices.

File: course_discovery/apps/course_metadata/models.py [hunk 8]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+    marketing_slug = models.CharField(
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py:526]: This value is soon to be unused by programs functionality in the LMS, so I'm conflicted about keeping it. Maybe it's good to keep around in case we need to link back out to the marketing site in the future?

File: course_discovery/apps/course_metadata/models.py [hunk 9]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+    # with related Courses are stored.
+    excluded_course_runs = models.ManyToManyField(CourseRun)
     partner = models.ForeignKey(Partner, null=True, blank=False)
+    overview = models.TextField(null=True, blank=True)
+    weeks_to_complete = models.PositiveSmallIntegerField(null=True, blank=True)
+    min_hours_effort_per_week = models.PositiveSmallIntegerField(null=True, blank=True)
+    max_hours_effort_per_week = models.PositiveSmallIntegerField(null=True, blank=True)
+    authoring_organizations = SortedManyToManyField(Organization, blank=True, related_name='authored_programs')
+
+    banner_image_url = models.URLField(null=True, blank=True, help_text=_('Image used atop marketing pages'))
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py]: This image is also used throughout programs pages in the LMS. Might be worth noting in the help text.

File: course_discovery/apps/course_metadata/models.py [hunk 10]
```diff
@@ -457,52 +473,70 @@ class Meta(object):
...
+    card_image_url = models.URLField(null=True, blank=True, help_text=_('Image used for discovery cards'))
+    video = models.ForeignKey(Video, default=None, null=True, blank=True)
+    expected_learning_items = SortedManyToManyField(ExpectedLearningItem, blank=True)
+    faq = SortedManyToManyField(FAQ, blank=True)
+
+    # MicroMasters-specific fields
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py]: Is there a need to mention MicroMasters here? To put this more generally, these - along with a few others, above - are really just optional fields. An open source user might choose to use any combination of them for their own ProgramType.

File: course_discovery/apps/course_metadata/models.py [hunk 11]
```diff
@@ -516,11 +550,52 @@ def marketing_url(self):
...
+    @property
+    def subjects(self):
+        subjects = [list(course.subjects.all()) for course in self.courses.all()]
+        subjects = itertools.chain.from_iterable(subjects)
+        return set(subjects)
+
+    @property
+    def price_ranges(self):
+        applicable_seat_types = self.type.applicable_seat_types.values_list('slug', flat=True)
+        data = Seat.objects.filter(course_run__in=self.course_runs, type__in=applicable_seat_types) \
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py]: The body of this method would be more readable if you name this variable after what it points to: `seats`.

File: course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+SEAT_TYPES = ('Audit', 'Credit', 'Professional', 'Verified',)
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py:6]: These names are ingrained in edx-platform, but over here in a newer project, should we be treating these the same way we treat "XSeries" and "MicroMasters"? Do we want to force every site operator to work with these seat types? Maybe we do, but I'm interested in what you think.

File: course_discovery/apps/api/serializers.py (hunk 1 above)
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: The challenge here is that https://github.com/edx/edx-mktg/pull/763 depends on this field. That being said, that work has not been deployed, so we can take this opportunity to change this to `card_image_url`.

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:500]: Users must explicitly specify the seat types that qualify for a program. No seats means no completion.

File: course_discovery/apps/course_metadata/models.py (hunk 5 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:519]: Maybe. That is somewhat out of scope.

File: course_discovery/apps/course_metadata/models.py (hunk 8 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:526]: This is necessary to present URLs for discovery cards.

File: course_discovery/apps/course_metadata/models.py (hunk 6 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:531]: The validation ensures our data remains clean, and we don't waste resources on unnecessary filtering or displaying invalid excluded course runs to users.

File: course_discovery/apps/course_metadata/models.py (hunk 3 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:564]: The focus of this property is to make later calculations simpler. If you need to do a grouping for API presentation, you should either adapt the output (e.g. group the data as you need) or create your own functionality in the API.

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (schenedx) [inline course_discovery/apps/course_metadata/models.py:500]: Oh.. That's so not what I thought. Glad I asked. Please add comment?

File: course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py (hunk 12 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py:6]: If others want to use different seat types/names, they are welcome to include them. We need some basic data to make the system work.

File: course_discovery/apps/course_metadata/admin.py [hunk 13]
```diff
@@ -36,10 +36,38 @@ class CourseRunAdmin(admin.ModelAdmin):
...
+    list_display = ('name',)
+
+
+@admin.register(SeatType)
+class SeatTypeAdmin(admin.ModelAdmin):
+    list_display = ('name', 'slug',)
+    readonly_fields = ('slug',)
+
+
+@admin.register(Endorsement)
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/admin.py:56]: Would it make more sense for Endorsements and CorporateEndorsements to be inline admin form within programs?

Comment (clintonb) [inline course_discovery/apps/course_metadata/admin.py:56]: Probably. We haven't really discussed administration, so I'd prefer to hold off on inlining forms until [ECOM-5009](https://openedx.atlassian.net/browse/ECOM-5009) is addressed.

File: course_discovery/apps/course_metadata/models.py (hunk 5 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:519]: This will be addressed in a follow-up PR.

Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:519]: https://openedx.atlassian.net/browse/ECOM-5173

File: course_discovery/apps/course_metadata/models.py (hunk 7 above)
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py]: @clintonb check out https://github.com/edx/course-discovery/pull/143#discussion_r69153015.

**Comment by clintonb** [comment]
@schenedx Data migration will happen as part of https://openedx.atlassian.net/browse/ECOM-5099.

**Comment by clintonb** [comment]
@edx/ecommerce all feedback addressed. Please take another look.

**Comment by schenedx** [comment]
👍

File: course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py:1]: Why are there two `course_metadata` migrations being introduced here?

Comment (rlucioni) [inline course_discovery/apps/course_metadata/migrations/0012_create_seat_types.py:1]: Ignore me, one is a data migration.

File: course_discovery/apps/course_metadata/models.py [hunk 15]
```diff
@@ -516,11 +561,52 @@ def marketing_url(self):
...
 
-        return None
+    @property
+    def languages(self):
+        return set([course_run.language for course_run in self.course_runs])
+
+    @property
+    def transcript_languages(self):
+        languages = [list(course_run.transcript_languages.all()) for course_run in self.course_runs]
+        languages = itertools.chain.from_iterable(languages)
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py:575]: Nice use of itertools.

File: course_discovery/apps/course_metadata/models.py [hunk 16]
```diff
@@ -516,11 +561,52 @@ def marketing_url(self):
...
+
+    @property
+    def price_ranges(self):
+        applicable_seat_types = self.type.applicable_seat_types.values_list('slug', flat=True)
+        seats = Seat.objects.filter(course_run__in=self.course_runs, type__in=applicable_seat_types) \
+            .values('currency') \
+            .annotate(models.Min('price'), models.Max('price'))
+        price_ranges = []
+
+        for datum in seats:
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/models.py]: Take this one all the way! `for seat in seats`...

**Comment by rlucioni** [comment]
👍
